expression/builtin_time.go

```diff
@@ -3280,1581 +3388,616 @@ func (du *baseDateArithmetical) vecGetIntervalFromReal(b *baseBuiltinFunc, input
 			baseBuiltinFunc:      bf,
 			baseDateArithmetical: newDateArithmeticalUtil(),
+			getInterval:          getIntervalFromDecimal,
+			vecGetInterval:       vecGetIntervalFromDecimal,
+			timeOp:               c.timeOp,
+			durationOp:           c.durationOp,
 		}
-		sig.setPbCode(tipb.ScalarFuncSig_AddDateDurationDecimal)
+		c.setPbCodeOp(sig, tipb.ScalarFuncSig_AddDateDurationDecimal, tipb.ScalarFuncSig_SubDateDurationDecimal)
 	}
 	return sig, nil
 }
 
-type builtinAddDateStringStringSig struct {
+type builtinAddSubDateAsStringSig struct {
 	baseBuiltinFunc
 	baseDateArithmetical
+	getDate        funcGetDateForDateAddSub
+	vecGetDate     funcVecGetDateForDateAddSub
+	getInterval    funcGetIntervalForDateAddSub
+	vecGetInterval funcVecGetIntervalForDateAddSub
+	timeOp         funcTimeOpForDateAddSub
 }
 
-func (b *builtinAddDateStringStringSig) Clone() builtinFunc {
-	newSig := &builtinAddDateStringStringSig{baseDateArithmetical: b.baseDateArithmetical}
+func (b *builtinAddSubDateAsStringSig) Clone() builtinFunc {
+	newSig := &builtinAddSubDateAsStringSig{
+		baseDateArithmetical: b.baseDateArithmetical,
+		getDate:              b.getDate,
+		vecGetDate:           b.vecGetDate,
+		getInterval:          b.getInterval,
+		vecGetInterval:       b.vecGetInterval,
+		timeOp:               b.timeOp,
+	}
 	newSig.cloneFrom(&b.baseBuiltinFunc)
 	return newSig
 }
 
-// evalTime evals ADDDATE(date,INTERVAL expr unit).
-// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_adddate
-func (b *builtinAddDateStringStringSig) evalTime(row chunk.Row) (types.Time, bool, error) {
+func (b *builtinAddSubDateAsStringSig) evalString(row chunk.Row) (string, bool, error) {
 	unit, isNull, err := b.args[2].EvalString(b.ctx, row)
 	if isNull || err != nil {
-		return types.ZeroTime, true, err
+		return types.ZeroTime.String(), true, err
 	}
 
-	date, isNull, err := b.getDateFromString(b.ctx, b.args, row, unit)
+	date, isNull, err := b.getDate(&b.baseDateArithmetical, b.ctx, b.args, row, unit)
 	if isNull || err != nil {
-		return types.ZeroTime, true, err
+		return types.ZeroTime.String(), true, err
 	}
 	if date.InvalidZero() {
-		return types.ZeroTime, true, handleInvalidTimeError(b.ctx, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, date.String()))
+		return types.ZeroTime.String(), true, handleInvalidTimeError(b.ctx, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, date.String()))
 	}
-	interval, isNull, err := b.getIntervalFromString(b.ctx, b.args, row, unit)
+
+	interval, isNull, err := b.getInterval(&b.baseDateArithmetical, b.ctx, b.args, row, unit)
 	if isNull || err != nil {
-		return types.ZeroTime, true, err
+		return types.ZeroTime.String(), true, err
 	}
 
-	result, isNull, err := b.add(b.ctx, date, interval, unit)
-	return result, isNull || err != nil, err
+	result, isNull, err := b.timeOp(&b.baseDateArithmetical, b.ctx, date, interval, unit, b.tp.GetDecimal())
+	if result.Microsecond() == 0 {
```

Comment:
Should it not check actual fsp, what if the result just happens to have a fsp > 0 but evaluates to 0?

Reply:
Bingo! This is the subtle part! I believe this is what MySQL does. From MySQL:
```
select date_add('2000-01-02 00:00:00.5', interval 500000 microsecond);
+----------------------------------------------------------------+
| date_add('2000-01-02 00:00:00.5', interval 500000 microsecond) |
+----------------------------------------------------------------+
| 2000-01-02 00:00:01                                            |
+----------------------------------------------------------------+
```

But yeah, you reminded me that I should've added this test case, even just for fun. I did this kind of tests several times when studying the essence behind all the MySQL madness. Sadly I forgot to add them into code.

Reply:
Case added.

Reply:
Thanks!